src/fsrs/algorithm.ts

```diff
@@ -275,4 +275,42 @@ export class FSRSAlgorithm {
   forgetting_curve(elapsed_days: number, stability: number): number {
     return +Math.pow(1 + (FACTOR * elapsed_days) / stability, DECAY).toFixed(8)
   }
+
+  /**
+   * Calculates the next state of memory based on the current state, time elapsed, and grade.
+   *
+   * @param memory_state - The current state of memory, which can be null.
+   * @param t - The time elapsed since the last review.
+   * @param {Rating} g Grade (Rating[0.Manual,1.Again,2.Hard,3.Good,4.Easy])
+   * @returns The next state of memory with updated difficulty and stability.
+   */
+  next_state(memory_state: FSRSState | null, t: number, g: number): FSRSState {
+    if (!memory_state) {
+      return {
+        difficulty: this.init_difficulty(g),
+        stability: this.init_stability(g),
+      }
+    }
+    const { difficulty: d, stability: s } = memory_state
+    if (g === 0) {
+      return {
+        difficulty: d,
+        stability: s,
+      }
+    }
+    const r = this.forgetting_curve(t, s)
+    const s_after_success = this.next_recall_stability(d, s, r, g)
+    const s_after_fail = this.next_forget_stability(d, s, r)
```

Comment:
https://github.com/open-spaced-repetition/ts-fsrs/blob/cdd9158eedf81f3b962bf63f8d49346fcdccf8e6/src/fsrs/impl/basic_scheduler.ts#L189-L201

Don't forget to clamp `s_after_fail` by `nextSMin`.